packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/apis/update.ts

```diff
@@ -32,84 +34,146 @@ export interface PerformUpdateParams<T = unknown> {
 }
 
 export const performUpdate = async <T>(
+  updateParams: PerformUpdateParams<T>,
+  apiContext: ApiExecutionContext
+): Promise<SavedObjectsUpdateResponse<T>> => {
+  const { type, id, options } = updateParams;
+  const { allowedTypes, helpers } = apiContext;
+  const namespace = helpers.common.getCurrentNamespace(options.namespace);
+
+  // check request is valid
+  const { validRequest, error } = isValidRequest({ allowedTypes, type, id });
+  if (!validRequest && error) {
+    throw error;
+  }
+
+  const maxAttempts = options.version ? 1 : 1 + DEFAULT_RETRY_COUNT;
+
+  // handle retryOnConflict manually by reattempting the operation in case of conflict errors
+  let response: SavedObjectsUpdateResponse<T>;
+  for (let currentAttempt = 1; currentAttempt <= maxAttempts; currentAttempt++) {
+    try {
+      response = await executeUpdate(updateParams, apiContext, { namespace });
+      break;
+    } catch (e) {
+      if (
+        SavedObjectsErrorHelpers.isConflictError(e) &&
+        e.retryableConflict &&
+        currentAttempt < maxAttempts
+      ) {
+        continue;
+      }
+      throw e;
+    }
+  }
+
+  return response!;
+};
+
+export const executeUpdate = async <T>(
   { id, type, attributes, options }: PerformUpdateParams<T>,
-  {
-    registry,
-    helpers,
-    allowedTypes,
-    client,
-    serializer,
-    migrator,
-    extensions = {},
-  }: ApiExecutionContext
+  { registry, helpers, client, serializer, extensions = {}, logger }: ApiExecutionContext,
+  { namespace }: { namespace: string | undefined }
 ): Promise<SavedObjectsUpdateResponse<T>> => {
   const {
     common: commonHelper,
     encryption: encryptionHelper,
     preflight: preflightHelper,
     migration: migrationHelper,
+    validation: validationHelper,
   } = helpers;
   const { securityExtension } = extensions;
 
-  const namespace = commonHelper.getCurrentNamespace(options.namespace);
-  if (!allowedTypes.includes(type)) {
-    throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
-  }
-  if (!id) {
-    throw SavedObjectsErrorHelpers.createBadRequestError('id cannot be empty'); // prevent potentially upserting a saved object with an empty ID
-  }
-
   const {
     version,
     references,
     upsert,
     refresh = DEFAULT_REFRESH_SETTING,
-    retryOnConflict = version ? 0 : DEFAULT_RETRY_COUNT,
+    migrationVersionCompatibility,
   } = options;
 
-  let preflightResult: PreflightCheckNamespacesResult | undefined;
-  if (registry.isMultiNamespace(type)) {
-    preflightResult = await preflightHelper.preflightCheckNamespaces({
-      type,
-      id,
-      namespace,
-    });
-  }
+  // Preflight calls to get the doc and check namespaces for multinamespace types.
+  const preflightDocResult = await preflightHelper.preflightGetDocForUpdate({
```

Comment:
I think we'll throw a GenericNotFound here if the proxy returns a 404 (see other comment)